### ISSUE TITLE
[Feature Request] Make OpenCore load Config.Plist via Relative Path

### ISSUE THREAD

**Issue by dakanji** [acidanthera/bugtracker#920]
Please consider making OpenCore load config.plist by considering a relative path rather than the absolute `EFI/OC/config.plist` that is currently the case.

Yes, the use case is to allow more than one instance of OpenCore easily and while it is understood you do not favour this but there is a valid reason that perhaps you might consider for those of us on Classic MacPros.

Basically, in other to apply vendor updates natively, VMM needs to be switched on but on day to day running, performance is better without this flag set.

Hence, many cMP users have two configs that are swapped at such times and it would be a lot easier to just have this set and ready to go in a way that can just be booted up.

It will make our lives a lot easier.

**Comment by vit9696** [comment]
I have already explained to you and other MacPro owners that you should switch to MacPro6,1 model and resolve the minor compatibility problem instead of messing with VMM. Since you insist on using VMM, well, you will suffer for your own good :-)

I do not believe there is a good way to implement what you ask mainly because bootstrap code will need to be aware of the OpenCore.efi and/or its configuration, but currently it only has the partition.

However, perhaps we can make an NVRAM variable containing a config.plist name/prefix/suffix(?). This could be `Security` → `AllowConfigOverride`. If it is set, and some variable is set, current configuration is unloaded and a custom configuration is loaded. @Download-Fritz what do you think? I am mainly thinking not about MP (these people never learn and ask about not-existent problems to be frank) but about testing OpenCore configuration on the same partition.

**Comment by dakanji** [comment]
> I am mainly thinking ... about testing OpenCore configuration on the same partition.

Well, now that you mention that. A friend of mine, John Smith, is a Classic MacPro user and has the following setup with multiple OpenCore Instances:
1. RELEASE version of current OpenCore Release with Hardware Acceleration in `EFI/OC`. Used for day to day boot of computer. 
2. DEBUG version of current OpenCore Release with Hardware Acceleration in `EFI/OC_DBG`. Sits there and only used whenever he needs a debug log.
3. A WIP OpenCore from GitHub in `EFI/OC_DEV`. This is built now and then to test things in advance of release. He migrates this to `EFI/OC` after each release by acidanthera and starts over.
4. OpenCore v0.5.5 with VMM setup for vendor updates in `EFI/OC_VMM`. He boots into this only when he needs to update macOS. He also maintains this as a recovery path into his Mac in case he messes something up. 
5. Release version of current OpenCore **without** Hardware Acceleration in `EFI/OC_ALT`. Used to get around issues on older versions on macOS with Acceleration. John is aware that [you have said this is not needed](https://github.com/acidanthera/bugtracker/issues/858#issuecomment-616142117) and max kernel etc can be used but doesn't yet know exactly how to implement this. He will retire this once he knows how. He saw [a post on the internet](https://forums.macrumors.com/threads/activate-amd-hardware-acceleration.2180095/post-28384969) with a tentative config but not yet able to swing it  

John does not boot directly into OpenCore and, erm, uses an alternative means when he boots to select from these options.

He patches the OpenCore code and builds for each variant himself to allow booting from the various `OC_XYZ` folders as they would otherwise pick up the config.plist in `EFI/OC`. 

He does not boot these using BOOT_X64 and uses the respective bootstrap.efi instead. 

It all works perfectly fine but he thought it would be easier and potential mix ups avoided if OpenCore.efi simply reads the config from the same folder it sits in.  He would then be able to use the same OpenCore.efi that are on the same release version in each folder, simply change the config as needed and not need to patch any more.

**Comment by vit9696** [comment]
The thing is, currently it is more complicated than you may have thought. When OpenCore starts it gets two parameters passed from the bootstrap:

- Bootloader filesystem root.
- Bootloader device path, optional.

The former is used to locate `EFI/OC/config.plist`. The latter is used to report OpenCore boot path to the operating system and mark the boot volume as hidden.

This approach is written with several things in mind including loading OpenCore from a trusted storage. For example, firmware (BIOS) or another device (USB flash booting config on the main drive) given that bootstrap code is written. If OpenCore starts to require using the same filesystem it is located on it may complicate the extension in the future.

**Comment by dakanji** [comment]
> The thing is, currently it is more complicated than you may have thought.

Noted. Thanks.

I am sure he will be ok as and his setup works as is and he has already written a script to automate building and patching.

**Comment by btwise** [comment]
> I am sure he will be ok as his setup works fine as is and he has already written a script to automate building and patching.

I need this method. Where can I get the modified code，THX

**Comment by dakanji** [comment]
> I need this method. Where can I get the modified code，THX

It needs some tweaks due to changes in v0.5.9. Will post a link later.

**Comment by dakanji** [comment]
@btwise 

https://gist.github.com/dakanji/9ff065f068ccb674ad96420e10430664.
As of May 2020, it builds all 0.5.x versions of OpenCore to date as well as the current development codebase.

**Comment by btwise** [comment]
> @btwise
> 
> See: https://gist.github.com/dakanji/be49c4e591f20ae5ef0701ca1392e621
> 
> As of May 2020, it builds all 0.5.x versions of OpenCore to date as well as the current development codebase.

I think you understand me wrong, I want to know how to use different config.plist to boot OC, not build script

**Comment by dakanji** [comment]
This is a feature request (This issue) that has not yet been decided on and it will presumably be decided by the dev team at some point whether to grant it or refuse it.

Either way, when they make a decision, you will know either way as this issue will be closed at the time.

In the meantime, you **have to build your own version** and patch it to run variants.

That's what the build script does. You have to read the issue posts to understand what's going on.

**Comment by hacker1024** [comment]
This change would be greatly appreciated, especially for times like these with a new macOS release right around the corner that may need different boot arguments and kexts.

At the moment, there's no easy way to prepare OC for a new macOS release with things that may break the existing installed one (resulting in the need to use the EFI shell or another OS to make changes instead).

**Comment by kommtzeitkonrad** [comment]
> At the moment, there's no easy way to prepare OC for a new macOS release with things that may break the existing installed one

How about this simple solution: Use an USB flash drive with OpenCore installed on it that is equipped with a config.plist that suits the needs of your alternate macOS installation. This way you can leave the EFI partition on your hard drive as it is.

**Comment by PRNDA** [comment]
> > At the moment, there's no easy way to prepare OC for a new macOS release with things that may break the existing installed one
> 
> Simple solution: Use an USB flash drive with OpenCore installed on it to boot from which is equipped with a `config.plist` that suits the needs of your alternate macOS installation. This way you can leave the EFI partition on your hard drive as it is.

Of course, I know this method, but what if you need 5 different profiles? What would you do, prepare 5 USB flash drives?

**Comment by mhaeuser** [comment]
> Of course, I know this method, but what if you need 5 different profiles? What would you do, prepare 5 USB flash drives?

Not necessarily, but at the very least 5 different partitions on the same drive, yes. While there absolutely can be sane ways to implement this, we mostly received borked and very edge-case use-cases, so this was not a priority. You make a theoretical point, but for it to be reasonable, you would actually need to provide 5 sane(!!) separate profile use-cases. :)

**Comment by jirrick** [comment]
I totally get that it's a super specific edge case, but here's one: You can have several different versions of MacOS on one drive for testing apps. A somehow recent Intel PC with AMD graphics can handle anything from High Sierra to Sequoia pretty well, but each different OS version needs a bit of tweaking, especially with SMBIOS and on the latest ones, more kexts. You can only do this right now by having lots of smaller disks each with its own EFI and switching them in the BIOS boot menu. It would be much better if they were all just partitions on a single 1TB SSD.

**Comment by mhaeuser** [comment]
@jirrick That works already. Relative paths are about multiple OpenCore instances sharing the same partition, not the same disk.

**Comment by jirrick** [comment]
> @jirrick That works already. Relative paths are about multiple OpenCore instances sharing the same partition, not the same disk.

Thanks for the reply. I actually spent this past weekend figuring out how to make everything work and I think I've cracked it (browsing any vaguely related forum was part of the journey). Here's a write up of my journey/solution: 

https://gist.github.com/jirrick/b7985532edf5ed92feffacb1f7345b40

**Comment by dakanji** [comment]
@jirrick ... I run 17 versions of Mac OS, 10.4 Tiger to 15.x Sequoia on two disks, on my MacPro3,1 and can boot them all via OpenCore; which sits in the ESP of third disk.

OpenCore could also have been on one of the two disks with the Mac OS installations, but I have learnt from experience that is best to have such boot loaders on the ESP of a separate disk that does not hold any OS. That way, if/when required due to whatever issues, I can pull the disk with OpenCore and still get into a supported Mac OS version to fix things.

This is possible because unlike back in the **Stone Age** when I made the original feature request here, or in the **Dark Age** when a subsequent related PR was rejected, https://github.com/acidanthera/OpenCorePkg/pull/131, your imagination is now the limit after the **Rennaisance** that followed. The current **post-Rennaisance** is not as great for various internal and external reasons, but we thankfully still have what went before. You can now run OpenCore out of multiple folders and are not limited to an `/EFI/OC` folder name.

So, all that is needed now is to create, in one ESP, as many OpenCore instance folders as required to target specific needs by amending the config file and kexts etc to match. Then use a separate boot manager, RefindPlus in my case but could be Apple's StartManager if available, to select which OpenCore instance I need. RefindPlus allows me to clearly label each instance.

You can use [MyBootMgr](https://forums.macrumors.com/threads/2231693) to set things up. It is aims to work OOB on Classic MacPro (3,1 to 5,1) and Xserve but, with manual intervention, can be used on any Intel Mac.

The basic arrangement I use is:
- `/EFI/OC_01`: Set up to boot Tiger to Catalina (`MyBootMgr Configured`)
- `/EFI/OC_02`: Set up to boot Big Sur and Monterey (`MyBootMgr Configured`)
- `/EFI/OC_03`: Set up to boot Ventura and Sonoma (`OCLP Configured`)
- `/EFI/OC_04`: Set up to boot Sequoia and Newer (`OCLP Configured - Needs Specific Boot Arg`)
- `/EFI/OC_05`: Set up to Install/Update Tiger to Monterey (`MyBootMgr Configured - Needs VMM switched on`)
- `/EFI/OC_06`: Set up to test stuff (`Typically copy in one of those above as needed`)
- `/EFI/OC_07`: Set up to hold a known working setup (`Typically a copy of OC_01 or OC_02`)

In summary, the chances of the original request being implemented are next to zero and the aim behind the original request, running different setups, can now be met fairly easily. The patching I used to need to do to the binary to allow this as mentioned in the referenced PR thread, is no longer needed today.

To conclude, leverage the folder naming flexibilty if this is important to you.

**Comment by mhaeuser** [comment]
@dakanji Thanks, I was about to look up sources…
@jirrick 
> It's like a restaurant menu where half the items will give you food poisoning, but there's no indication which ones. Classic.

Gotta love LLM humour…

**Comment by jirrick** [comment]
@dakanji thanks for extensive summary. I've basically figured out very similar solution and written small summary for future reference (linked in my previous post).

Basically what made it work well for me was figuring out how to use the `InstanceIdentifier` property together with `.contentVisibility` files to limit which macOS volumes are visible from each instance so only the compatible ones are shown.

Excuse some of the childish humor in there.

**Comment by dakanji** [comment]
> Basically what made it work well for me was figuring out how to use the InstanceIdentifier property together with .contentVisibility files to limit which macOS volumes are visible from each instance so only the compatible ones are shown.

If only you knew what battles took place to get that in ... Still dealing with the trauma!

**Comment by btwise** [comment]
For the function of using multiple configuration files, I realized it like this:
In the OcMain stage, before reading the default config.plist, if there are multiple valid plist files under EFI/OC, a profile selector menu will appear, so that you can use any profile to be started for subsequent boot, instead of just looking for config.plist. I have tested that it can be used!

`EFI_STATUS
OcMiscEarlyInit (
  IN  OC_STORAGE_CONTEXT  *Storage,
  OUT OC_GLOBAL_CONFIG    *Config,
  IN  OC_RSA_PUBLIC_KEY   *VaultKey  OPTIONAL
  )
{
  EFI_STATUS      Status;
  CHAR8           *ConfigData;
  UINT32          ConfigDataSize;
  EFI_TIME        BootTime;
  CONST CHAR8     *AsciiVault;
  OCS_VAULT_MODE  Vault;
  UINTN           PciDeviceInfoSize;
  CONST CHAR16    *ConfigPath;

  //
  //  Smart configuration file selector
  // If multiple valid configs exist, show selection menu
  // If only one valid config exists, use it directly
  //
  ConfigPath = OcSelectConfigurationFile (Storage, VaultKey);

  ConfigData = OcStorageReadFileUnicode (
                 Storage,
                 ConfigPath,
                 &ConfigDataSize
                 );

  if (ConfigData != NULL) {
    DEBUG ((DEBUG_INFO, "OC: Loaded configuration of %u bytes\n", ConfigDataSize));

    Status = OcConfigurationInit (Config, ConfigData, ConfigDataSize, NULL);
    if (EFI_ERROR (Status)) {
      DEBUG ((DEBUG_ERROR, "OC: Failed to parse configuration!\n"));
      CpuDeadLoop ();
      return EFI_UNSUPPORTED; ///< Should be unreachable.
    }

    FreePool (ConfigData);
  } else {
    DEBUG ((DEBUG_ERROR, "OC: Failed to load configuration!\n"));
    CpuDeadLoop ();
    return EFI_UNSUPPORTED; ///< Should be unreachable.
  }`
